---
title: "debugging_dag"
output: github_document
date: "2024-01-31"
---
Install packages
```{r message=FALSE, warning=FALSE}
library(tibble)
library(simulateGP)
library(dplyr)
library(TwoSampleMR)
library(parallel)
library(data.table)
library(ggplot2)
```

```{r}

gx_to_gp <- function(gx, h2x, prev){
  x_prime <- qnorm(prev, 0, 1)
  p <- pnorm(x_prime, mean=-gx, sd=sqrt(1-h2x), lower.tail = TRUE)
  return(p)
}
logistic_assoc <- function(y, x)
{
	logistic_model <- summary(glm(y ~ x, family="binomial"))$coefficients
	return(list(
		ahat=logistic_model[1,1],
		bhat=logistic_model[2,1],
		se=logistic_model[2,2],
		fval=logistic_model[2,3]^2,
		pval=logistic_model[2,4],
		n=length(y)
	))
}
dgmodel <- function(nid, #making large sample size so enough individuals for case control analysis
  h2_known, #h-squared value (form of r-squared) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4667278/
  h2_unknown, #to total 0.25 of total H-squared (form of r-squared) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4667278/
  d_prev, #glioma prevalence https://www.ncbi.nlm.nih.gov/books/NBK441874/#:~:text=In%20the%20United%20States%2C%20there,the%20least%20malignant%20brain%20tumors.
  #b_lr0=rev_mr_result[row,]$beta,  #adding rev MR results into data generating model
  gr_maf, #need to estimate this allele frequency
  rsq_gr0,   
  rsq_gc0, 
  rsq_lr0,
  b_u1r1, 
  b_u1l, 
  b_u2c1, 
  b_u2l, 
  b_dr1, #non causal marker increases beta 1 in disease
  b_dc1, #causal marker increases beta 1 in disease
  b_c0c1, 
  b_r0r1, 
  b_u2c0, 
  b_u1r0, 
  b_u1d, 
  b_u2d, 
  b_gcc0,
  b_gc0l,#set to this as beta of c0 is this 
  var_l, #estimated variance of liability to be 1
  vp, 
  nsnp)  #assuming variance of genetic liability is 1
{
  u1 <- rnorm(nid) 
  u2 <- rnorm(nid) 
gwashits <- tribble(
    ~beta, ~af,
    0.26, 0.54
)
gc <- sapply(gwashits$af, \(x) rbinom(nid, 2, x)) #why not rbinom(nid, 2, gwashits$af) the latter creates a matrix

c0 <- gc * b_gcc0 + u2*b_u2c0 + rnorm(nid, sd=sqrt(1-b_gcc0^2-b_u2c0^2))

#gr_maf <- runif(nsnp, 0.01, 0.99)
gprs_maf <- rep(gr_maf, nsnp)

prs_b <- rnorm(nsnp)#randomly generated betas

gprs <- make_geno(nid, nsnp, gprs_maf)

prs_known <- (gprs %*% prs_b) %>% {scale(.) * sqrt(h2_known)}

prs_unknown <- rnorm(nid, 0, sd = sqrt(h2_unknown))

prs <- prs_known + prs_unknown

rsq_prs <- h2_known - (0.1)^2 #check this where 0.1 comes from

l <- c0 * b_gc0l + scale(prs) * sqrt(rsq_prs)+ u2 * b_u2l + rnorm(nid, 0, sd = 1 - sqrt(rsq_prs + b_gc0l^2 + b_u2l^2))

gr <- rbinom(nid, 2, gr_maf) #find out if gr_maf should be changed to it's own one

r0 <- scale(gr) * sqrt(rsq_gr0) + scale(l) * sqrt(rsq_lr0) + u1 * b_u1r0 + rnorm(nid, sd=sqrt(1 - rsq_gr0 - rsq_lr0 - b_u1r0^2))

prob_l <- gx_to_gp(gx=scale(l), h2x=rsq_prs + b_gc0l^2 + b_u1d^2 + b_u2d^2, prev = d_prev)  

d <- rbinom(nid, 1, prob_l) 

c1 <- c0 * b_c0c1 + u2 * b_u2c1 + d * b_dc1 

r1 <- r0 * b_r0r1 + u1 * b_u1r1 + d * b_dr1 

join_mat <- cbind(gprs, gr, gc)   

phen <- tibble(
  u1, u2, r0, r1, c0, c1, prs, l, prob_l, d
)  
return(list(geno_gc=gc,geno_prs=gprs, geno_gr=gr, phen=phen, geno_join=join_mat, af=gprs_maf))}  

dgmodel_check <- function(dat)
{
  print(tibble(
    col = names(dat$phen),
    vars = apply(dat$phen, 2, var), 
    means = colMeans(dat$phen)
  ), n=25)
  print(cor(dat$phen$prs, dat$phen$l)^2)
}
dgmodel_analysis <- function(dat, ncase, ncontrol, protein_gwas, rev_mr_result, row, iteration)
{
  phen <- dat$phen
  res_all <- bind_rows( 
    fast_assoc(y=phen$d, x=phen$prs) %>% as_tibble() %>% mutate(x="prs", y="d"),
    fast_assoc(y=phen$c0, x=phen$d) %>% as_tibble() %>% mutate(x="c0", y="d"),
    fast_assoc(y=phen$r0, x=phen$d) %>% as_tibble() %>% mutate(x="r0", y="d"),
    fast_assoc(y=phen$prs, x=phen$c0) %>% as_tibble() %>% mutate(x="prs", y="c0"),
    fast_assoc(y=phen$prs, x=phen$r0) %>% as_tibble() %>% mutate(x="prs", y="r0"),
    fast_assoc(y=phen$c1, x=phen$d) %>% as_tibble() %>% mutate(x="c1", y="d"),
    fast_assoc(y=phen$r1, x=phen$d) %>% as_tibble() %>% mutate(x="r1", y="d"),
    fast_assoc(y=phen$prs, x=phen$c1) %>% as_tibble() %>% mutate(x="prs", y="c1"),
    fast_assoc(y=phen$prs, x=phen$c1) %>% as_tibble() %>% mutate(x="prs", y="r1")
  ) %>%
    mutate(study="all")
  
  obs <- bind_rows(phen[phen$d == 0,][1:ncontrol,], phen[phen$d == 1,][1:ncase,])  
  case <- length(subset(obs$d, obs$d==1))
  print("Number of cases: ") 
  print(case)
  control <- length(subset(obs$d, obs$d==0))
  print("Number of controls: ")
  print(control)
  res_obs <- bind_rows(
    fast_assoc(y=obs$d, x=obs$prs) %>% as_tibble() %>% mutate(x="prs", y="d"),
    fast_assoc(y=obs$c0, x=obs$d) %>% as_tibble() %>% mutate(x="c0", y="d"),
    fast_assoc(y=obs$r0, x=obs$d) %>% as_tibble() %>% mutate(x="r0", y="d"),
    fast_assoc(y=obs$prs, x=obs$c0) %>% as_tibble() %>% mutate(x="prs", y="c0"),
    fast_assoc(y=obs$prs, x=obs$r0) %>% as_tibble() %>% mutate(x="prs", y="r0"),
    fast_assoc(y=obs$c1, x=obs$d) %>% as_tibble() %>% mutate(x="c1", y="d"),
    fast_assoc(y=obs$r1, x=obs$d) %>% as_tibble() %>% mutate(x="r1", y="d"),
    fast_assoc(y=obs$prs, x=obs$c1) %>% as_tibble() %>% mutate(x="prs", y="c1"),
    fast_assoc(y=obs$prs, x=obs$r1) %>% as_tibble() %>% mutate(x="prs", y="r1"),
  )  %>%
    mutate(study="observational")

  prot <- phen[phen$d==0,][1:protein_gwas, ] 
  res_protein <- bind_rows(
    fast_assoc(y=prot$d, x=prot$prs) %>% as_tibble() %>% mutate(x="prs", y="d"),
    fast_assoc(y=prot$c0, x=prot$d) %>% as_tibble() %>% mutate(x="c0", y="d"),
    fast_assoc(y=prot$r0, x=prot$d) %>% as_tibble() %>% mutate(x="r0", y="d"),
    fast_assoc(y=prot$prs, x=prot$c0) %>% as_tibble() %>% mutate(x="prs", y="c0"),
    fast_assoc(y=prot$prs, x=prot$r0) %>% as_tibble() %>% mutate(x="prs", y="r0"),
    fast_assoc(y=prot$c1, x=prot$d) %>% as_tibble() %>% mutate(x="c1", y="d"),
    fast_assoc(y=prot$r1, x=prot$d) %>% as_tibble() %>% mutate(x="r1", y="d"),
    fast_assoc(y=prot$prs, x=prot$c1) %>% as_tibble() %>% mutate(x="prs", y="c1"),
    fast_assoc(y=prot$prs, x=prot$r1) %>% as_tibble() %>% mutate(x="prs", y="r1"),
  ) %>%
    mutate(study="protein_gwas")
  out_all <- paste0(rev_mr_result[row,]$prot, iteration, "_res_all.csv")
  write.csv(res_all, file = out_all, row.names = FALSE, quote = FALSE)
  out_obs <- paste0(rev_mr_result[row,]$prot, iteration, "_res_obs.csv")
  write.csv(res_obs, file = out_obs, row.names = FALSE, quote = FALSE)
  out_protein <- paste0(rev_mr_result[row,]$prot, iteration,"_res_protein.csv")
  write.csv(res_protein, file = out_protein, row.names = FALSE, quote = FALSE)
  return(bind_rows(obs%>%
    mutate(study="observational"), prot%>%
    mutate(study="protein_gwas")))
} 

ivw_analysis <- function(dat){
  phen <- dat$phen
  gen_gr <- dat$geno_gr
mr_dat <- get_effs(phen$prs, phen$`r0[, 1]`, gen_gr) #had to set r0[,1] so that it is pulling first column 
print("Reverse MR")
mr(mr_dat, metho=c("mr_ivw")) %>% str()
gen_gc <- dat$geno_gc
mr_dat <- get_effs(phen$`c1[, 1]`, phen$d, gen_gc) 
print("Forward MR")
mr(mr_dat, metho=c("mr_ivw", "mr_wald_ratio")) %>% str()

print("2SLS")
summary(systemfit::systemfit(phen$prs ~ phen$r0, method="2SLS", inst = ~gen)) 
print("Observational")
summary(lm(phen$prs ~ phen$r0))
}
```

Reverse MR
```{r}
rev_mr_prot <- function(dat, n_protein_gwas, ncontrol, ncase){
  phen <- dat$phen
  prot <- phen[1:n_protein_gwas, ] #first group of protein individuals prot <- phen[phen$d==0,][1:n_protein_gwas, ]
  ind <- do.call(paste0, dat$phen) %in% do.call(paste0, prot)
  genosubset <- dat$geno_prs[ind, ]
  geno_mat <- data.matrix(genosubset)
  phenosubset <- dat$phen[ind, ]
  stopifnot(nrow(genosubset) == nrow(phenosubset))
  y <- phenosubset$r0
  g <- geno_mat
  gwasy <- gwas(y,g, logistic = FALSE)
  
  no_prot_phen <- dat$phen[!ind,] #remove protein individuals so no overlap
  no_prot_gen <- data.matrix(dat$geno_prs[!ind,])
  obs <- bind_rows(no_prot_phen[no_prot_phen$d == 0,][1:ncontrol,], no_prot_phen[no_prot_phen$d == 1,][1:ncase,])
  ind_obs <- do.call(paste0, no_prot_phen) %in% do.call(paste0, obs)
  phenosubset <- no_prot_phen[ind_obs, ]
  genosubset <- no_prot_gen[ind_obs, ]
  geno_mat <- data.matrix(genosubset)
  stopifnot(nrow(genosubset) == nrow(phenosubset))
  x <- phenosubset$d #change this to d and do logistic regression to check this as it's binary, l for fast assoc
  g <- geno_mat
  gwasx <- gwas(x,g, logistic = TRUE)
  harmonise <- merge_exp_out(gwasx, gwasy, "X", "Y")
  prs_snp <- subset(harmonise, harmonise$pval.exposure<5e-8)#select SNPs reaching threshold
  print(prs_snp)
  if (dim(prs_snp)[1]>0){
   mr_res <- mr(prs_snp, metho=c("mr_ivw", "mr_wald_ratio")) 
  all_mr_res <- mr(harmonise, metho=c("mr_ivw", "mr_wald_ratio"))
   return(list(pval=mr_res, all=all_mr_res))
  }else{
    mr_res <- tribble(
  ~id.exposure, ~id.outcome, ~outcome, ~exposure, ~method, ~nsnp,~b, ~se, ~pval, ~mr,
 " ", " ", " ", " "," "," ", " ", " ", " ","rev",
)
all_mr_res <- mr(harmonise, metho=c("mr_ivw", "mr_wald_ratio"))
return(c(pval=mr_res, all=all_mr_res))
  }}
```
Nested case control analysis
```{r}
nested_case_control <- function(dat, nestedcase, nestedcontrol){
    phen <- dat$phen
 nested <- bind_rows(phen[phen$d == 0,][1:nestedcase,], phen[phen$d == 1,][1:nestedcontrol,]) #487 example based on nested case control study
  log_nested<- logistic_assoc(y=nested$d, x=nested$r0) %>% as_tibble() %>% mutate(x="r0", y="d") #case control estimate 
  #fast_nested<- fast_assoc(y=nested$d, x=nested$r0) %>% as_tibble() %>% mutate(x="r0", y="d") #case control estimate 
  return(c(casecontrol=log_nested))} #fast_cc=fast_nested
```

FORWARD MR
```{r}
fwd_mr_prot <- function(dat, n_protein_gwas, ncontrol, ncase){
  phen <- dat$phen
  prot <- phen[1:n_protein_gwas, ] #first group of protein individuals prot <- phen[phen$d==0,][1:n_protein_gwas, ]
  ind <- do.call(paste0, dat$phen) %in% do.call(paste0, prot)
  genosubset <- dat$geno_join[ind, ]
  geno_mat <- data.matrix(genosubset)
  phenosubset <- dat$phen[ind, ]
  stopifnot(nrow(genosubset) == nrow(phenosubset))
  x <- phenosubset$c0
  g <- geno_mat
  gwasx <- gwas(x,g, logistic = FALSE)
  
  no_prot_phen <- dat$phen[!ind,] #remove protein individuals so no overlap
  no_prot_gen <- data.matrix(dat$geno_join[!ind,])
  obs <- bind_rows(no_prot_phen[no_prot_phen$d == 0,][1:ncontrol,], no_prot_phen[no_prot_phen$d == 1,][1:ncase,])
  ind_obs <- do.call(paste0, no_prot_phen) %in% do.call(paste0, obs)
  phenosubset <- no_prot_phen[ind_obs, ]
  genosubset <- no_prot_gen[ind_obs, ]
  geno_mat <- data.matrix(genosubset)
  stopifnot(nrow(genosubset) == nrow(phenosubset))
  y <- phenosubset$d #change this to d and do logistic regression to check this as it's binary
  g <- geno_mat
    gwas_dat <- data.frame()
 gwasy <- gwas(y,g, logistic = TRUE)
  harmonise <- merge_exp_out(gwasx, gwasy, "X", "Y")
  c_snp <- harmonise[101,] #select simulated causal SNP
  mr_res <- mr(c_snp, metho=c("mr_ivw", "mr_wald_ratio"))
  return(mr_res)}
```

```{r}
all_assoc <- function(dat){
y <- dat$phen$r0
  g <- dat$geno_prs
  gwasy <- gwas(y,g, logistic = FALSE)
  x <- dat$phen$d #change this to d and do logistic regression to check this as it's binary, l for fast assoc
  g <- dat$geno_prs
gwasx <- gwas(x,g, logistic = TRUE)
  harmonise <- merge_exp_out(gwasx, gwasy, "X", "Y")
  prs_snp <- subset(harmonise, harmonise$pval.exposure<5e-8)#select SNPs reaching threshold
   mr_res <- mr(prs_snp, metho=c("mr_ivw", "mr_wald_ratio")) 
  all_mr_res <- mr(harmonise, metho=c("mr_ivw", "mr_wald_ratio"))
  return(c(real_pval=mr_res, real=all_mr_res))
  }
```


```{r message=TRUE}
#calculate r-squared
rev_mr_result <- tribble(
  ~beta, ~prot, ~rsq_lr0, ~se,
 0.0879946, "HMGCS1", 0.0052923, 0.0210034, 
 #0.0832134, "ATAD2", 0.0049975, 0.0204426,
 #0.0829339, "BMPER", 0.0049653, 0.0204403, 
 #0.0827007, "IL5", 0.0049354, 0.0204447,
 #0.0826643, "IL23R", 0.0049313, 0.0204444,
 #0.0497226, "BCAN",0.0016068,0.0067785 #ukbb n=33436
 #0.0447559, "SPINK2",0.0013336,0.0067233 #ukbb n=33187
 #0.01168, "test egfr", 0.00863,0.00218, 
)
#fval <- (0.0880/0.0210)^2
#rsq <- fval/(3301-2+fval)
#fval <- (beta/se)^2
#rsq <- fval/(n-2+fval)
param  <- expand.grid(
  n_prot=34557,#c(34557, 35559, 3301),
  af = seq(0.2, 0.4, by=0.1),
  b=c(0.0879946, -0.0354690,0.0497226), 
  rsq_lr0=c(0.0052923,0.0008683,0.0016068),
  nestedcasecontrol=487,#c(113, 487,593),
  sims=1:2
)
sims <- lapply(1:nrow(param), function(i){
  print(param[i,])
dat <- dgmodel(nid=60000, #making large sample size so enough individuals for case control analysis
  h2_known=0.2, #h-squared value (form of r-squared) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4667278/ - changing the h2_known as we have a large sample size to accommodate case control data needed for glioma but this inturn weakens the instruemnts so we increase h2_known
  h2_unknown=0.19, #to total 0.25 of total H-squared (form of r-squared) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4667278/
  d_prev=4000/10000, #glioma prevalence https://www.ncbi.nlm.nih.gov/books/NBK441874/#:~:text=In%20the%20United%20States%2C%20there,the%20least%20malignant%20brain%20tumors.
  #b_lr0=rev_mr_result[row,]$beta,  #adding rev MR results into data generating model
  gr_maf=param$af[i], #need to estimate this allele frequency 0.1*i
  rsq_gr0=0,   
  rsq_gc0=0, 
  rsq_lr0=param$rsq_lr0[i],
  b_u1r1=0, 
  b_u1l=0, 
  b_u2c1=0, 
  b_u2l=0, 
  b_dr1=0, #non causal marker stays the same in disease
  b_dc1=1, #causal marker increases beta 1 in disease
  b_c0c1=1, 
  b_r0r1=1, 
  b_u2c0=0, 
  b_u1r0=0, 
  b_u1d=0, 
  b_u2d=0, 
  b_gcc0=0.13, #not 0.13 which is used to be as need to make 0.26 for causal effect
  b_gc0l=0, #set to this as beta of c0 is this 
  var_l=1, #estimated variance of liability to be 1
  vp=1, #estimated variance of protein to be 1
  nsnp=99) 
#dgmodel_check(dat)
#studies <- dgmodel_analysis(dat, ncase=12496, ncontrol=18190, protein_gwas = 34557, rev_mr_result, row, iteration=i)
#print("studies")
rev <- rev_mr_prot(dat, n_protein_gwas=param$n_prot[i], ncontrol=18190, ncase=12496)
rev$all$mr <- "rev_all"
rev$pval$mr <- "rev_pval"
nest <- nested_case_control(dat, nestedcase=param$nestedcasecontrol[i], nestedcontrol=param$nestedcasecontrol[i])
rev$casecontrol <- nest
fwd <- fwd_mr_prot(dat, n_protein_gwas=param$n_prot[i], ncontrol=18190, ncase=12496)
fwd$mr <- "fwd"
rev$fwd <- fwd
real <- all_assoc(dat)
rev$real <- real
rev$af <- param$af[i]
rev$b <- param$b[i]
rev$n <- param$n[i]
return(rev)
}) 
```

```{r}
sims <- lapply(1:10, function(i){
dat <- dgmodel(nid=60000, #making large sample size so enough individuals for case control analysis
  h2_known=0.15, #h-squared value (form of r-squared) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4667278/ - changing the h2_known as we have a large sample size to accommodate case control data needed for glioma but this inturn weakens the instruemnts so we increase h2_known
  h2_unknown=0.19, #to total 0.25 of total H-squared (form of r-squared) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4667278/
  d_prev=4000/10000, #glioma prevalence https://www.ncbi.nlm.nih.gov/books/NBK441874/#:~:text=In%20the%20United%20States%2C%20there,the%20least%20malignant%20brain%20tumors.
  #b_lr0=rev_mr_result[row,]$beta,  #adding rev MR results into data generating model
  gr_maf=0.3, #need to estimate this allele frequency 0.1*i
  rsq_gr0=0,   
  rsq_gc0=0, 
  rsq_lr0=0.0052923,
  b_u1r1=0, 
  b_u1l=0, 
  b_u2c1=0, 
  b_u2l=0, 
  b_dr1=0, #non causal marker stays the same in disease
  b_dc1=1, #causal marker increases beta 1 in disease
  b_c0c1=1, 
  b_r0r1=1, 
  b_u2c0=0, 
  b_u1r0=0, 
  b_u1d=0, 
  b_u2d=0, 
  b_gcc0=0.13,
  b_gc0l=0, #set to this as beta of c0 is this 
  var_l=1, #estimated variance of liability to be 1
  vp=1, #estimated variance of protein to be 1
  nsnp=99) 
rev <- rev_mr_prot(dat, n_protein_gwas=34557, ncontrol=18190, ncase=12496)
rev$all$mr <- "rev_all"
rev$pval$mr <- "rev_pval"
nest <- nested_case_control(dat, nestedcase=487, nestedcontrol=487)
rev$casecontrol <- nest
fwd <- fwd_mr_prot(dat, n_protein_gwas=34557, ncontrol=18190, ncase=12496)
fwd$mr <- "fwd"
rev$fwd <- fwd
real <- all_assoc(dat)
rev$real <- real
rev$af <- param$af[i]
rev$b <- param$b[i]
rev$n <- param$n[i]
return(rev)
}) 

real_mr <- data.table()
for (i in 1:10){
    row <- sims[[i]]$real
    real_mr <- rbind(real_mr, row)
}

res <- data.frame(type = c(rep("revmr_all", nrow(real_mr)), rep("actual",nrow(real_mr))), b=c(real_mr$real.b, rep(0.0879946, nrow(real_mr))), run=c(1:nrow(real_mr), 1:nrow(real_mr)), se=c(real_mr$real.se, rep(0.0210034, nrow(real_mr))))
ggplot(res, aes(x=run, y=b, col=type))+geom_point()+geom_line()+ylim(NA, rev_mr_result$beta+rev_mr_result$se+0.05)+geom_errorbar(aes(ymin = b-se, ymax = b+se), width=0.1) 

rev_mr_all <- data.table()
for (i in 1:10){
  row <- sims[[i]]$all
  rev_mr_all <- rbind(rev_mr_all, row, fill=TRUE)
}
nested <- data.table()
for (i in 1:10){
  row <- sims[[i]]$casecontrol
  nested <- rbind(nested, row, fill=TRUE)
}
res <- data.frame(type = c(rep("revmr_all", nrow(real_mr)), rep("revmr_selected",nrow(rev_mr_all)),rep("actual",nrow(real_mr)), rep("nested", nrow(nested)))
, b=c(real_mr$real.b,rev_mr_all$b, rep(0.0879946, nrow(real_mr)), nested$casecontrol.bhat), run=c(1:nrow(rev_mr_all), 1:nrow(real_mr),1:nrow(real_mr), 1:nrow(nested)), se=c(real_mr$real.se, rev_mr_all$se,rep(0.0210034, nrow(real_mr)), nested$casecontrol.se))
ggplot(res, aes(x=run, y=b, col=type))+geom_point()+geom_line()+ylim(NA, rev_mr_result$beta+rev_mr_result$se+0.05)+geom_errorbar(aes(ymin = b-se, ymax = b+se), width=0.1) +geom_hline(yintercept=mean(real_mr$real.b), color = "darkorchid")+geom_hline(yintercept=mean(rev_mr_all$b), color = "darkcyan")+geom_hline(yintercept=mean(nested$casecontrol.bhat), color = "darkolivegreen4")
```

```{r}
fwd_mr <- data.table()
for (i in 1:nrow(param)){
  row <- sims[[i]]$fwd
  fwd_mr <- rbind(fwd_mr, row)
}
rev_mr_all <- data.table()
for (i in 1:nrow(param)){
  row <- sims[[i]]$all
  rev_mr_all <- rbind(rev_mr_all, row)
}
rev_mr_pval <- data.table()
for (i in 1:nrow(param)){
  row <- sims[[i]]$pval
  rev_mr_pval <- rbind(rev_mr_pval, row)
}
casecontrol_log <- data.table()
for (i in 1:nrow(param)){
  row <- sims[[i]]$casecontrol
  casecontrol_log <- rbind(casecontrol_log, row)
}
af <- data.table()
for (i in 1:nrow(param)){
  row <- sims[[i]]$af
  avg <- mean(row)
  af <- rbind(af, avg)
}
real_mr <- data.table()
 for (i in 1:nrow(param)){
     row <- sims[[i]]$real
     real_mr <- rbind(real_mr, row)
 }
real_mr_all <- real_mr[,10:18]
```

Power and bias
```{r}
rev_mr_all$af <- af
rev_mr_all$n_prot <- param$n_prot
rev_mr_all$beta <- param$b
real_mr_all$af <- af
real_mr_all$n_prot <- param$n_prot
real_mr_all$beta <- param$b
real_mr_all$mr <- "real"
colnames(real_mr_all) <- c("id.exposure","id.outcome","outcome","exposure","method","nsnp","b", "se", "pval", "af", "n_prot", "beta", "mr")
casecontrol_log$af <- af
casecontrol_log$n_prot <- param$n_prot
casecontrol_log$beta <- param$b
res <- rbind(rev_mr_all, real_mr_all)
res <- res[,7:13]
cc_res <- casecontrol_log[,c(2,3,5,9,10,11)]
colnames(cc_res) <- c("b", "se", "pval", "af", "n_prot", "beta")
cc_res$mr <- "casecontrol"
res <- rbind(cc_res, res)

sum_res <- res %>% group_by(n_prot, af, beta, mr) %>%
        summarise(bias=mean((b)-beta),
        power=sum(pval<0.05/n())) #divided by the group size to adjust for multiple testing

ggplot(sum_res, aes(x=af, y=bias, color=mr)) +
    geom_point() +
    geom_line() +
    facet_grid(n_prot ~beta)

ggplot(sum_res, aes(x=af, y=power, color=mr)) +
    geom_point() +
    geom_line() +
    facet_grid(n_prot ~ beta)
```






Data visualization - all rev mr
```{r}
for (file in 1:nrow(rev_mr_result)){
res <- data.frame(type = c(rep("revmr", reps), rep("n_cc", reps), rep("pval_revmr", reps),rep("actual",reps)), b=c(rev_mr_all$b, casecontrol$bhat, rev_mr_pval$b,rep(rev_mr_result$beta, reps)), run=c(1:reps, 1:reps, 1:reps, 1:reps), se=c(rev_mr_all$se, casecontrol$se, rev_mr_pval$se,rep(rev_mr_result$se, reps)))
ggplot(res, aes(x=run, y=b, col=type))+geom_point()+geom_line()+ylim(NA, rev_mr_result$beta+rev_mr_result$se+0.05)+geom_errorbar(aes(ymin = b-se, ymax = b+se), width=0.1) 
ggsave(paste0("interval_output/", rev_mr_result$prot[file], ".pdf"))}
```

Analysis
```{r}

all_res <- sims$mr
fwd_mr_res <- subset(all_res, all_res$mr=="fwd")
rev_mr_res <- subset(all_res, all_res$mr=="rev_all")#change this to pval if needed
rev_mr_res_pval <- subset(all_res, all_res$mr=="rev_pval")
#plot <- fwd_mr_res[,c(7,8)]
#pairs(plot, panel=function(x,y){
#  points(x,y)
#  abline(lm(y~x), col="red")})

for (row in 1:nrow(rev_mr_res)){
  if (rev_mr_res[row,1]== " "){
    rev_mr_res <- rev_mr_res[-row,]
  }
  else{}}
na.omit(rev_mr_res)
fwd_mr_res$b <- as.numeric(fwd_mr_res$b)
var_mean_fwd <-  bind_rows(
    vars = var(fwd_mr_res$b), 
    means = mean(fwd_mr_res$b),
    type="fwd")
rev_mr_res$b <- as.numeric(rev_mr_res$b)
var_mean_rev <-  bind_rows(
    vars = var(rev_mr_res$b), 
    means = mean(rev_mr_res$b),
    type="rev")
total <- rbind(var_mean_fwd, var_mean_rev)
hist(fwd_mr_res$b)
points(x=1,y=1, col="red", pch=15)
boxplot(fwd_mr_res$b)
hist(rev_mr_res$b)
points(x=0.08,y=1, col="red", pch=15)
boxplot(rev_mr_res$b)

```
observational and all information on r0
```{r}
non_causal_all <- data.frame()
for (prot in rev_mr_result$prot){
  for (sim in 1:reps){
  file <- read.csv(paste0(prot,sim, "_res_all.csv"))
  non_casual <- subset(file, file$x=="r0" & file$y=="d")
  non_causal_all <- rbind(non_causal_all, non_casual %>%
    mutate(protein=prot))
}}
```

```{r}
comb <- data.frame(all_res$cc.bhat, non_causal_all$bhat, rev_mr_res$b, rev_mr_res_pval$b)
```




Plot associations between variables in disease and non disease state
```{r}
new_obs <- bind_rows(phen[phen$d == 0,][1:100,], phen[phen$d == 1,][1:100,])
obs_dis <- subset(new_obs, new_obs$d ==1)
obs_nodis <- subset(new_obs, new_obs$d ==0)
plot(obs_nodis, col="black")
plot(obs_dis, col="red")
```

Case control MR
```{r}
#Reverse MR
genosubset <- dat$geno_gr[ind, ]
geno_mat <- data.matrix(genosubset)
phenosubset <- dat$phen[ind, ]
stopifnot(nrow(genosubset) == nrow(phenosubset))
mr_dat <- get_effs(phenosubset$`r0[, 1]`, phenosubset$d, geno_mat)
print("Reverse MR")
mr(mr_dat, metho=c("mr_ivw", "mr_wald_ratio")) %>% str()
```
 


TO DO:
- Run scenarios
- Is reverse MR more or less sensitive to disease than case control/prospective studies
- To do reverse MR with protein data - use SNP exposure from one population for exposure estimate and then get SNP outcome estimate from other population
- Look at sensitivity and specificity to predict non causal proteins - area under curve 
What condition is revMR better than case/control or prospective
- Lose SNPs in gr0 compared to gc0 - protein have 99 and disease has only 1

Create function to take R0 data from protein cohort and d variable from GWAS data from different subset of population and perform reverse MR


Obs vs expected:
```{r}
expected_vs_observed_replication <- function(b_disc, b_rep, se_disc, se_rep, alpha) { 
  p_sign <- pnorm(-abs(b_disc) / se_disc) * pnorm(-abs(b_disc) / se_rep) + ((1 - pnorm(-abs(b_disc) / se_disc)) * (1 - pnorm(-abs(b_disc) / se_rep))) 
  p_sig <- pnorm(-abs(b_disc) / se_rep + qnorm(alpha / 2)) + (1 - pnorm(-abs(b_disc) / se_rep - qnorm(alpha / 2)))
  p_rep <- pnorm(abs(b_rep)/se_rep, lower.tail=FALSE) 
  res <- tibble::tibble( 
    nsnp=length(b_disc), 
    metric=c("Sign", "Sign", "P-value", "P-value"), 
    datum=c("Expected", "Observed", "Expected", "Observed"), 
    value=c(sum(p_sign, na.rm=TRUE), sum(sign(b_disc) == sign(b_rep)), sum(p_sig, na.rm=TRUE), sum(p_rep < alpha, na.rm=TRUE)) 
  ) 
  return(list(res=res, variants=dplyr::tibble(sig=p_sig, sign=p_sign))) 
} 
expected_vs_observed_replication(rev_mr_result$beta, rev_mr_result$se, rev_mr_comb$b, rev_mr_comb$se, alpha = 0.05) 
expected_vs_observed_replication(rev_mr_result$beta, rev_mr_result$se, non_causal_all$bhat, non_causal_all$se, alpha = 0.05)
expected_vs_observed_replication(rev_mr_result$beta, rev_mr_result$se, non_causal_obs$bhat, non_causal_obs$se, alpha = 0.05)
```


Sensitivity and specificity
```{r}

```

Reverse MR
```{r}
rev_mr_prot <- function(dat, n_protein_gwas, ncontrol, ncase, nsnp){
  phen <- dat$phen
  prot <- bind_rows(phen[phen$d==0,][1:n_protein_gwas, ], phen[phen$d == 1,][1:100,])#first 3301
  ind <- do.call(paste0, dat$phen) %in% do.call(paste0, prot)
  genosubset <- dat$geno_join[ind, ]
  geno_mat <- data.matrix(genosubset)
  phenosubset <- dat$phen[ind, ]
  stopifnot(nrow(genosubset) == nrow(phenosubset))
  y <- phenosubset$c0
  g <- geno_mat
  gwas_dat <- data.frame()
  for (i in 1:ncol(g)){
    gwas_dat <- rbind(gwas_dat, simulateGP::fast_assoc(y,g[,i]))
  }
  out_dat <- data.frame()
  for (y in 1:nrow(gwas_dat)){
    prot_dat <- dplyr::tibble(
      SNP= y, 
      outcome= "Y",
      id.outcome= "Y",
      beta.outcome=gwas_dat[y,]$bhat,
      se.outcome=gwas_dat[y,]$se,
      pval.outcome=gwas_dat[y,]$pval,
      samplesize.outcome=gwas_dat[y,]$n,
      units.outcome="SD",
      rsq.outcome=gwas_dat[y,]$fval/(gwas_dat[y,]$fval+gwas_dat[y,]$n-2) #n-2 for degrees of freedom https://analystprep.com/cfa-level-1-exam/quantitative-methods/coefficient-of-determination-and-f-statistic/
    )
    out_dat<- rbind(out_dat, prot_dat)
  }
  
  no_prot_phen <- dat$phen[!ind,] #remove protein individuals so no overlap
  no_prot_gen <- data.matrix(dat$geno_join[!ind,])
  obs <- bind_rows(no_prot_phen[no_prot_phen$d == 0,][1:ncontrol,], no_prot_phen[no_prot_phen$d == 1,][1:ncase,])
  ind_obs <- do.call(paste0, no_prot_phen) %in% do.call(paste0, obs)
  phenosubset <- no_prot_phen[ind_obs, ]
  genosubset <- no_prot_gen[ind_obs, ]
  geno_mat <- data.matrix(genosubset)
  stopifnot(nrow(genosubset) == nrow(phenosubset))
  x <- phenosubset$d #change this to d and do logistic regression to check this as it's binary, l for fast assoc
  g <- geno_mat
  #gwas_dat <- data.frame()
#  for (i in 1:ncol(g)){
#    gwas_dat <- rbind(gwas_dat, simulateGP::fast_assoc(x,g[,i]))
#  }
gwas_dat <- data.frame()
for (i in 1:ncol(g)){
    gwas_dat <- rbind(gwas_dat, logistic_assoc(x,g[,i]))
}
exp_dat <- data.frame()
for (x in 1:nrow(gwas_dat)){
    cc_dat <- dplyr::tibble(
      SNP= x,
      exposure= "X",
      id.exposure= "X",
      beta.exposure=gwas_dat[x,]$bhat,
      se.exposure=gwas_dat[x,]$se,
      pval.exposure=gwas_dat[x,]$pval,
      samplesize.exposure=gwas_dat[x,]$n,
      units.exposure="SD",
      rsq.exposure=gwas_dat[x,]$fval/(gwas_dat[x,]$fval+gwas_dat[x,]$n-2) #n-2 for degrees of freedom https://analystprep.com/cfa-level-1-exam/quantitative-methods/coefficient-of-determination-and-f-statistic/
    )
    exp_dat<- rbind(exp_dat, cc_dat)
  }
  harmonise <- merge(exp_dat, out_dat, by="SNP")
  harmonise$mr_keep <- TRUE
  n_snp <- harmonise[1:nsnp,]
  prs_snp <- subset(n_snp, n_snp$pval.exposure<0.05)#select SNPs reaching threshold
  print(prs_snp)
  if (dim(prs_snp)[1]>0){
   mr_res <- mr(prs_snp, metho=c("mr_ivw", "mr_wald_ratio")) 
  all_mr_res <- mr(n_snp, metho=c("mr_ivw", "mr_wald_ratio"))
  nested <- bind_rows(obs[obs$d == 0,][1:ncontrol,], obs[obs$d == 1,][1:100,])
  nested<- fast_assoc(y=nested$d, x=nested$r0) %>% as_tibble() %>% mutate(x="prs", y="d") #case control estimate 
   return(list(pval=mr_res, all=all_mr_res, casecontrol=nested))
  }else{
    mr_res <- tribble(
  ~id.exposure, ~id.outcome, ~outcome, ~exposure, ~method, ~nsnp,~b, ~se, ~pval, ~mr,
 " ", " ", " ", " "," "," ", " ", " ", " ","rev",
)
all_mr_res <- mr(n_snp, metho=c("mr_ivw", "mr_wald_ratio"))
cc<- fast_assoc(y=phenosubset$d, x=phenosubset$r0) %>% as_tibble() %>% mutate(x="prs", y="d") #case control estimate 
return(c(pval=mr_res, all=all_mr_res, casecontrol=cc))
  }}
```

```{r}
fwd_mr <- data.table()
for (i in 1:reps){
  row <- sims[[i]]$fwd
  fwd_mr <- rbind(fwd_mr, row)
}
rev_mr_all <- data.table()
for (i in 1:reps){
  row <- sims[[i]]$all
  rev_mr_all <- rbind(rev_mr_all, row)
}
rev_mr_pval <- data.table()
for (i in 1:reps){
  row <- sims[[i]]$pval
  rev_mr_pval <- rbind(rev_mr_pval, row)
}
casecontrol <- data.table()
for (i in 1:reps){
  row <- sims[[i]]$casecontrol
  casecontrol <- rbind(casecontrol, row)
}
betas <- data.table(revmr_all = rev_mr_all$b, revmr_pval= rev_mr_pval$b, cc=casecontrol$bhat)
```

Analysis
```{r}

all_res <- sims$mr
fwd_mr_res <- subset(all_res, all_res$mr=="fwd")
rev_mr_res <- subset(all_res, all_res$mr=="rev_all")#change this to pval if needed
rev_mr_res_pval <- subset(all_res, all_res$mr=="rev_pval")
#plot <- fwd_mr_res[,c(7,8)]
#pairs(plot, panel=function(x,y){
#  points(x,y)
#  abline(lm(y~x), col="red")})

for (row in 1:nrow(rev_mr_res)){
  if (rev_mr_res[row,1]== " "){
    rev_mr_res <- rev_mr_res[-row,]
  }
  else{}}
na.omit(rev_mr_res)
fwd_mr_res$b <- as.numeric(fwd_mr_res$b)
var_mean_fwd <-  bind_rows(
    vars = var(fwd_mr_res$b), 
    means = mean(fwd_mr_res$b),
    type="fwd")
rev_mr_res$b <- as.numeric(rev_mr_res$b)
var_mean_rev <-  bind_rows(
    vars = var(rev_mr_res$b), 
    means = mean(rev_mr_res$b),
    type="rev")
total <- rbind(var_mean_fwd, var_mean_rev)
hist(fwd_mr_res$b)
points(x=1,y=1, col="red", pch=15)
boxplot(fwd_mr_res$b)
hist(rev_mr_res$b)
points(x=0.08,y=1, col="red", pch=15)
boxplot(rev_mr_res$b)

```
observational and all information on r0
```{r}
non_causal_all <- data.frame()
for (prot in rev_mr_result$prot){
  for (sim in 1:reps){
  file <- read.csv(paste0(prot,sim, "_res_all.csv"))
  non_casual <- subset(file, file$x=="r0" & file$y=="d")
  non_causal_all <- rbind(non_causal_all, non_casual %>%
    mutate(protein=prot))
}}
```

```{r}
comb <- data.frame(all_res$cc.bhat, non_causal_all$bhat, rev_mr_res$b, rev_mr_res_pval$b)
```




Plot associations between variables in disease and non disease state
```{r}
new_obs <- bind_rows(phen[phen$d == 0,][1:100,], phen[phen$d == 1,][1:100,])
obs_dis <- subset(new_obs, new_obs$d ==1)
obs_nodis <- subset(new_obs, new_obs$d ==0)
plot(obs_nodis, col="black")
plot(obs_dis, col="red")
```

Case control MR
```{r}
#Reverse MR
genosubset <- dat$geno_gr[ind, ]
geno_mat <- data.matrix(genosubset)
phenosubset <- dat$phen[ind, ]
stopifnot(nrow(genosubset) == nrow(phenosubset))
mr_dat <- get_effs(phenosubset$`r0[, 1]`, phenosubset$d, geno_mat)
print("Reverse MR")
mr(mr_dat, metho=c("mr_ivw", "mr_wald_ratio")) %>% str()
```
 


TO DO:
- Run scenarios
- Is reverse MR more or less sensitive to disease than case control/prospective studies
- To do reverse MR with protein data - use SNP exposure from one population for exposure estimate and then get SNP outcome estimate from other population
- Look at sensitivity and specificity to predict non causal proteins - area under curve 
What condition is revMR better than case/control or prospective
- Lose SNPs in gr0 compared to gc0 - protein have 99 and disease has only 1

Create function to take R0 data from protein cohort and d variable from GWAS data from different subset of population and perform reverse MR


Obs vs expected:
```{r}
expected_vs_observed_replication <- function(b_disc, b_rep, se_disc, se_rep, alpha) { 
  p_sign <- pnorm(-abs(b_disc) / se_disc) * pnorm(-abs(b_disc) / se_rep) + ((1 - pnorm(-abs(b_disc) / se_disc)) * (1 - pnorm(-abs(b_disc) / se_rep))) 
  p_sig <- pnorm(-abs(b_disc) / se_rep + qnorm(alpha / 2)) + (1 - pnorm(-abs(b_disc) / se_rep - qnorm(alpha / 2)))
  p_rep <- pnorm(abs(b_rep)/se_rep, lower.tail=FALSE) 
  res <- tibble::tibble( 
    nsnp=length(b_disc), 
    metric=c("Sign", "Sign", "P-value", "P-value"), 
    datum=c("Expected", "Observed", "Expected", "Observed"), 
    value=c(sum(p_sign, na.rm=TRUE), sum(sign(b_disc) == sign(b_rep)), sum(p_sig, na.rm=TRUE), sum(p_rep < alpha, na.rm=TRUE)) 
  ) 
  return(list(res=res, variants=dplyr::tibble(sig=p_sig, sign=p_sign))) 
} 
expected_vs_observed_replication(rev_mr_result$beta, rev_mr_result$se, rev_mr_comb$b, rev_mr_comb$se, alpha = 0.05) 
expected_vs_observed_replication(rev_mr_result$beta, rev_mr_result$se, non_causal_all$bhat, non_causal_all$se, alpha = 0.05)
expected_vs_observed_replication(rev_mr_result$beta, rev_mr_result$se, non_causal_obs$bhat, non_causal_obs$se, alpha = 0.05)
```


Sensitivity and specificity
```{r}

```
