---
title: "debugging_dag"
output: github_document
date: "2024-01-31"
---
Install packages
```{r message=FALSE, warning=FALSE}
library(tibble)
library(simulateGP)
library(dplyr)
library(TwoSampleMR)
```

```{r}
gx_to_gp <- function(gx, h2x, prev){
  x_prime <- qnorm(prev, 0, 1)
  p <- pnorm(x_prime, mean=-gx, sd=sqrt(1-h2x), lower.tail = TRUE)
  return(p)
}
dgmodel <- function(nid, #making large sample size so enough individuals for case control analysis
  h2_known, #h-squared value (form of r-squared) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4667278/
  h2_unknown, #to total 0.25 of total H-squared (form of r-squared) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4667278/
  d_prev, #glioma prevalence https://www.ncbi.nlm.nih.gov/books/NBK441874/#:~:text=In%20the%20United%20States%2C%20there,the%20least%20malignant%20brain%20tumors.
  #b_lr0=rev_mr_result[row,]$beta,  #adding rev MR results into data generating model
  gr_maf, #need to estimate this allele frequency
  rsq_gr0,   
  rsq_gc0, 
  rsq_lr0,
  b_u1r1, 
  b_u1l, 
  b_u2c1, 
  b_u2l, 
  b_dr1, #non causal marker increases beta 1 in disease
  b_dc1, #causal marker increases beta 1 in disease
  b_c0c1, 
  b_r0r1, 
  b_u2c0, 
  b_u1r0, 
  b_u1d, 
  b_u2d, 
  b_gcc0,
  b_gc0l,#set to this as beta of c0 is this 
  var_l, #estimated variance of liability to be 1
  vp, nsnp)  #assuming variance of genetic liability is 1
{
  u1 <- rnorm(nid) 
  u2 <- rnorm(nid) 
gwashits <- tribble(
    ~beta, ~af,
    0.26, 0.54
)
 gc <- sapply(gwashits$af, \(x) rbinom(nid, 2, x)) 


c0 <- gc * b_gcc0 + u2*b_u2c0 + rnorm(nid, sd=sqrt(1-b_gcc0^2-b_u2c0^2))

rsq_prs <- h2_known - (0.1)^2
prs <- rnorm(nid, mean=0, sd=sqrt(vp*h2_known)) 

l <- c0*b_gc0l +prs*sqrt(rsq_prs)+u2*b_u2l
gr <- make_geno(nid, nsnp, gr_maf)
l_for_r0 <- matrix(l,             
               nrow = nrow(l),
               ncol = ncol(gr))
  r0 <- gr * sqrt(rsq_gr0) + l_for_r0 * sqrt(rsq_lr0) + u1 * b_u1r0 + rnorm(nid, sd=sqrt(1-rsq_gr0-rsq_lr0^2-b_u1r0^2)) 
  prob_l <- gx_to_gp(gx=scale(l), h2x=rsq_prs + b_gc0l^2 + b_u1d^2 + b_u2d^2, prev = d_prev) 
 
  d <- rbinom(nid, 1, prob_l) 

  c1 <-c0 *b_c0c1 + u2 * b_u2c1 + d * b_dc1 + rnorm(nid, sd=sqrt(1 - b_u2c1^2- b_c0c1^2))
  r1 <-  r0 * b_r0r1 + u1 * b_u1r1 + d * b_dr1 + rnorm(nid, sd=sqrt(1-b_r0r1^2-b_u1r1^2)) 
   phen <- tibble(
    u1, u2, r0[,1], r1[,1], c0[,1], c1[,1], prs, l, prob_l, d
  )
  return(list(geno_gc=gc,geno_gr=gr, phen=phen))} 

dgmodel_check <- function(dat)
{
  print(tibble(
    col = names(dat$phen),
    vars = apply(dat$phen, 2, var), 
    means = colMeans(dat$phen)
  ), n=25)
  print(cor(dat$phen$prs, dat$phen$l)^2)
}
dgmodel_analysis <- function(dat, ncase, ncontrol, protein_gwas, rev_mr_result)
{
  phen <- dat$phen
  res_all <- bind_rows( 
    fast_assoc(y=phen$d, x=phen$prs) %>% as_tibble() %>% mutate(x="prs", y="d"),
    fast_assoc(y=phen$`c0[, 1]`, x=phen$d) %>% as_tibble() %>% mutate(x="c0_1", y="d"),
    fast_assoc(y=phen$`r0[, 1]`, x=phen$d) %>% as_tibble() %>% mutate(x="r0", y="d"),
    fast_assoc(y=phen$prs, x=phen$`c0[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="c0_1"),
    fast_assoc(y=phen$prs, x=phen$`r0[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="r0"),
    fast_assoc(y=phen$`c1[, 1]`, x=phen$d) %>% as_tibble() %>% mutate(x="c1_1", y="d"),
    fast_assoc(y=phen$`r1[, 1]`, x=phen$d) %>% as_tibble() %>% mutate(x="r1", y="d"),
    fast_assoc(y=phen$prs, x=phen$`c1[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="c1_1"),
    fast_assoc(y=phen$prs, x=phen$`r1[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="r1")
  ) %>%
    mutate(study="all")
  
  obs <- bind_rows(phen[phen$d == 0,][1:ncontrol,], phen[phen$d == 1,][1:ncase,])  
  case <- length(subset(obs$d, obs$d==1))
  paste("Number of cases: ", case) 
  control <- length(subset(obs$d, obs$d==0))
  paste("Number of controls: ", control)
  res_obs <- bind_rows(
    fast_assoc(y=obs$d, x=obs$prs) %>% as_tibble() %>% mutate(x="prs", y="d"),
    fast_assoc(y=obs$`c0[, 1]`, x=obs$d) %>% as_tibble() %>% mutate(x="c0_1", y="d"),
    fast_assoc(y=obs$`r0[, 1]`, x=obs$d) %>% as_tibble() %>% mutate(x="r0", y="d"),
    fast_assoc(y=obs$prs, x=obs$`c0[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="c0_1"),
    fast_assoc(y=obs$prs, x=obs$`r0[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="r0"),
    fast_assoc(y=obs$`c1[, 1]`, x=obs$d) %>% as_tibble() %>% mutate(x="c1_1", y="d"),
    fast_assoc(y=obs$`r1[, 1]`, x=obs$d) %>% as_tibble() %>% mutate(x="r1", y="d"),
    fast_assoc(y=obs$prs, x=obs$`c1[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="c1_1"),
    fast_assoc(y=obs$prs, x=obs$`r1[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="r1"),
  )  %>%
    mutate(study="observational")

  prot <- phen[phen$d==0,][1:protein_gwas, ] 
  res_protein <- bind_rows(
    fast_assoc(y=prot$d, x=prot$prs) %>% as_tibble() %>% mutate(x="prs", y="d"),
    fast_assoc(y=prot$`c0[, 1]`, x=prot$d) %>% as_tibble() %>% mutate(x="c0_1", y="d"),
    fast_assoc(y=prot$`r0[, 1]`, x=prot$d) %>% as_tibble() %>% mutate(x="r0", y="d"),
    fast_assoc(y=prot$prs, x=prot$`c0[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="c0_1"),
    fast_assoc(y=prot$prs, x=prot$`r0[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="r0"),
    fast_assoc(y=prot$`c1[, 1]`, x=prot$d) %>% as_tibble() %>% mutate(x="c1_1", y="d"),
    fast_assoc(y=prot$`r1[, 1]`, x=prot$d) %>% as_tibble() %>% mutate(x="r1", y="d"),
    fast_assoc(y=prot$prs, x=prot$`c1[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="c1_1"),
    fast_assoc(y=prot$prs, x=prot$`r1[, 1]`) %>% as_tibble() %>% mutate(x="prs", y="r1"),
  ) %>%
    mutate(study="protein_gwas")
  out_all <- paste0(rev_mr_result[row,]$prot, "_res_all.csv")
  write.csv(res_all, file = out_all, row.names = FALSE, quote = FALSE)
  out_obs <- paste0(rev_mr_result[row,]$prot, "_res_obs.csv")
  write.csv(res_obs, file = out_obs, row.names = FALSE, quote = FALSE)
  out_protein <- paste0(rev_mr_result[row,]$prot, "_res_protein.csv")
  write.csv(res_protein, file = out_protein, row.names = FALSE, quote = FALSE)
  return(bind_rows(obs%>%
    mutate(study="observational"), prot%>%
    mutate(study="protein_gwas")))
} 


ivw_analysis <- function(dat){
  phen <- dat$phen
  gen_gr <- dat$geno_gr
mr_dat <- get_effs(phen$prs, phen$`r0[, 1]`, gen_gr) #had to set r0[,1] so that it is pulling first column 
print("Reverse MR")
mr(mr_dat, metho=c("mr_ivw")) %>% str()
gen_gc <- dat$geno_gc
mr_dat <- get_effs(phen$`c1[, 1]`, phen$d, gen_gc) 
print("Forward MR")
mr(mr_dat, metho=c("mr_ivw", "mr_wald_ratio")) %>% str()

print("2SLS")
summary(systemfit::systemfit(phen$prs ~ phen$r0, method="2SLS", inst = ~gen)) 
print("Observational")
summary(lm(phen$prs ~ phen$r0))
}
```

```{r}
rev_mr_result <- tribble(
  ~beta, ~prot, ~rsq_lr0, ~se,
 0.0880, "HMGCS1", 0.2040, 0.0210,
 0.0832, "ATAD2", 0.1824, 0.0204,
 0.0829, "BMPER", 0.1811, 0.0204,
 0.0827, "IL5", 0.1802, 0.0204,
 0.0827, "IL23R", 0.1802, 0.0204,
)
rev_mr_comb <- data.frame()
fwd_mr_comb <- data.frame()
for(row in 1:nrow(rev_mr_result)){
dat <- dgmodel(nid=100000, #making large sample size so enough individuals for case control analysis
  h2_known=0.06, #h-squared value (form of r-squared) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4667278/
  h2_unknown=0.19, #to total 0.25 of total H-squared (form of r-squared) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4667278/
  d_prev=6/100000, #glioma prevalence https://www.ncbi.nlm.nih.gov/books/NBK441874/#:~:text=In%20the%20United%20States%2C%20there,the%20least%20malignant%20brain%20tumors.
  #b_lr0=rev_mr_result[row,]$beta,  #adding rev MR results into data generating model
  gr_maf=0.20, #need to estimate this allele frequency
  rsq_gr0=0,   
  rsq_gc0=0, 
  rsq_lr0=rev_mr_result[row,]$rsq_lr0,
  b_u1r1=0, 
  b_u1l=0, 
  b_u2c1=0, 
  b_u2l=0, 
  b_dr1=0, #non causal marker stays the same in disease
  b_dc1=1, #causal marker increases beta 1 in disease
  b_c0c1=1, 
  b_r0r1=1, 
  b_u2c0=0, 
  b_u1r0=0, 
  b_u1d=0, 
  b_u2d=0, 
  b_gcc0=0.13,
  b_gc0l=0.13, #set to this as beta of c0 is this 
  var_l=1, #estimated variance of liability to be 1
  vp=1, 
  nsnp=99) #estimated variance of protein to be 1

dgmodel_check(dat)

studies <- dgmodel_analysis(dat, ncase=12496, ncontrol=18190, protein_gwas = 3301, rev_mr_result)

rev <- rev_mr_prot(dat, n_protein_gwas=3301, ncontrol=18190, ncase=12496)
rev_mr_comb <- rbind(rev_mr_comb, rev)
obs <- subset(studies, studies$study == "observational")
obs <- obs[, -11] #obs without identifier
fwd <- fwd_mr_cc(dat, obs)
fwd_mr_comb <- rbind(fwd_mr_comb, fwd)
}
#ivw_analysis(dat)

```

Plot associations between variables in disease and non disease state
```{r}
new_obs <- bind_rows(phen[phen$d == 0,][1:100,], phen[phen$d == 1,][1:100,])
obs_dis <- subset(new_obs, new_obs$d ==1)
obs_nodis <- subset(new_obs, new_obs$d ==0)
plot(obs_nodis, col="black")
plot(obs_dis, col="red")
```

Case control MR
```{r}
#Reverse MR
genosubset <- dat$geno_gr[ind, ]
geno_mat <- data.matrix(genosubset)
phenosubset <- dat$phen[ind, ]
stopifnot(nrow(genosubset) == nrow(phenosubset))
mr_dat <- get_effs(phenosubset$`r0[, 1]`, phenosubset$d, geno_mat)
print("Reverse MR")
mr(mr_dat, metho=c("mr_ivw", "mr_wald_ratio")) %>% str()
```
 


TO DO:
- Run scenarios
- Is reverse MR more or less sensitive to disease than case control/prospective studies
- To do reverse MR with protein data - use SNP exposure from one population for exposure estimate and then get SNP outcome estimate from other population
- Look at sensitivity and specificity to predict non causal proteins - area under curve 
What condition is revMR better than case/control or prospective
- Lose SNPs in gr0 compared to gc0 - protein have 99 and disease has only 1

Create function to take R0 data from protein cohort and d variable from GWAS data from different subset of population and perform reverse MR
```{r}
rev_mr_prot <- function(dat, n_protein_gwas, ncontrol, ncase){
phen <- dat$phen
gen_gr <- dat$geno_gr
prot <- phen[phen$d==0,][1:n_protein_gwas, ] #first 3301
ind <- do.call(paste0, dat$phen) %in% do.call(paste0, prot)
genosubset <- dat$geno_gr[ind, ]
geno_mat <- data.matrix(genosubset)
phenosubset <- dat$phen[ind, ]
stopifnot(nrow(genosubset) == nrow(phenosubset))
x <- phenosubset$`r0[, 1]`
g <- geno_mat
gwas_dat <- data.frame()
for (i in 1:ncol(g)){
gwas_dat <- rbind(gwas_dat, simulateGP::fast_assoc(x,g[,i]))
}
exp_dat <- data.frame()
for (x in 1:nrow(gwas_dat)){
  prot_dat <- dplyr::tibble(
  SNP= x, 
  exposure= "X",
  id.exposure= "X",
  beta.exposure=gwas_dat[x,]$bhat,
  se.exposure=gwas_dat[x,]$se,
  pval.exposure=gwas_dat[x,]$pval,
  samplesize.exposure=gwas_dat[x,]$n,
  units.exposure="SD",
  rsq.exposure=gwas_dat[x,]$fval/(gwas_dat[x,]$fval+gwas_dat[x,]$n-2) #n-2 for degrees of freedom https://analystprep.com/cfa-level-1-exam/quantitative-methods/coefficient-of-determination-and-f-statistic/
)
exp_dat<- rbind(exp_dat, prot_dat)
}

no_prot_phen <- dat$phen[!ind,] #remove protein individuals so no overlap
no_prot_gc <- data.matrix(dat$geno_gc[!ind,])
obs <- bind_rows(no_prot_phen[no_prot_phen$d == 0,][1:ncontrol,], no_prot_phen[no_prot_phen$d == 1,][1:ncase,])
ind_obs <- do.call(paste0, no_prot_phen) %in% do.call(paste0, obs)
phenosubset <- no_prot_phen[ind_obs, ]
genosubset <- no_prot_gc[ind_obs, ]
geno_mat <- data.matrix(genosubset)
stopifnot(nrow(genosubset) == nrow(phenosubset))
y <- phenosubset$d
g <- geno_mat
gwas_dat <- data.frame()
for (i in 1:ncol(g)){
gwas_dat <- rbind(gwas_dat, simulateGP::fast_assoc(y,g[,i]))
}
out_dat <- dplyr::tibble(
  SNP= 1:nrow(gwas_dat),#sort this 
  outcome= "Y",
  id.outcome= "Y",
  beta.outcome=gwas_dat$bhat,
  se.outcome=gwas_dat$se,
  pval.outcome=gwas_dat$pval,
  samplesize.outcome=gwas_dat$n,
  units.outcome="SD",
  rsq.outcome=gwas_dat$fval/(gwas_dat$fval+gwas_dat$n-2) #n-2 for degrees of freedom https://analystprep.com/cfa-level-1-exam/quantitative-methods/coefficient-of-determination-and-f-statistic/
)
harmonise <- merge(exp_dat, out_dat, by="SNP")
harmonise$mr_keep <- TRUE
mr_res <- mr(harmonise, metho=c("mr_ivw", "mr_wald_ratio"))
return(mr_res)}

```

Case control forward MR
```{r}
##Forward MR
fwd_mr_cc <- function(dat, obs){
ind <- do.call(paste0, dat$phen) %in% do.call(paste0, obs)
genosubset <- dat$geno_gc[ind, ]
geno_mat <- data.matrix(genosubset)
phenosubset <- dat$phen[ind, ]
stopifnot(nrow(genosubset) == nrow(phenosubset))
mr_dat <- get_effs(phenosubset$`c1[, 1]`, phenosubset$d, geno_mat)
print("Forward MR")
mr_res <- mr(mr_dat, metho=c("mr_ivw", "mr_wald_ratio"))
return(mr_res)}
```

observational and all information on r0
```{r}
non_causal_obs <- data.frame()
for (prot in rev_mr_result$prot){
  file <- read.csv(paste0(prot, "_res_obs.csv"))
  non_casual <- subset(file, file$x=="r0" & file$y=="d")
  non_causal_obs <- rbind(non_causal_obs, non_casual%>%
    mutate(protein=prot))
}

non_causal_all <- data.frame()
for (prot in rev_mr_result$prot){
  file <- read.csv(paste0(prot, "_res_all.csv"))
  non_casual <- subset(file, file$x=="r0" & file$y=="d")
  non_causal_all <- rbind(non_causal_all, non_casual %>%
    mutate(protein=prot))
}
```



Sensitivity and specificity
```{r}

```

